### DISCUSSION TITLE
Project Planning

### DISCUSSION THREAD

**Discussion by coltonehrman** [Software-Developer-Academy/mentor-match#4] [General]
Project planning and requirements discussion.

Template:

```
## Project Goals

1. ...

## Success Metrics 

1. ...

## Features

### Feature 1

#### Actions

**Comment by coltonehrman** [comment]
## Project Goals

1. Enable efficient and meaningful connections between mentors/mentees
2. Manage connections and sessions between mentors/mentees
3. Facilitate sessions between mentors/mentees
4. Promote mentors to mentees based on good fit

## Success Metrics 

1. DAU/MAU
2. New account signups
3. Positive member feedback

## Features

### Actors

1. Mentee
2. Mentor

### Accounts

#### Pages

1. Signup Page
3. Signin Page
4. Account Dashboard

#### Shared

Shared functions across account types

1. Create member account
1. Assign roles to account (ie: mentee, mentor)
1. Assign skills to account (ie: js, node.js, react)
1. Assign interests to account (ie: tech, books, nature)
5. Assign time-zone to account
6. Assign availability preferences to account

### Search

#### Pages

1. Search w/ filters

#### Components

1. Popular mentors

#### CTA

1. Connect (to connect a mentor <-> mentee)

#### Mentee

1. Search for mentors
3. Filters mentors based on attributes
4. Find popular mentors
6. Connect with mentor
7. Schedule session with mentor
8. Provide feedback to mentor

#### Mentor

1. Search for mentees
3. Filters mentees based on attributes
4. Connect with mentee
5. Schedule session with mentor
7. Provide feedback to mentee

### Profile View

A public view of member profile for another actor to interact with.

#### Pages

1. Profile Detail

#### Components

1. Name
3. Avatar / image
4. Intro
5. Skills
6. Interests
7. Roles

#### CTA

1. Connect (to connect a mentor <-> mentee)
2. Schedule session

**Reply by tjay22** [reply]
@coltonehrman I elaborated these in the user journeys below. We can further break down features from there.

**Reply by coltonehrman** [reply]
@tjay22 Perfect, thank you! It looks great :)

**Reply by brucesailes** [reply]
Hello everyone!

**Reply by tjay22** [reply]
Hi Bruce!

**Reply by brucesailes** [reply]
How can I help you guys? I’m at work currently but I can pitch in once I get off.

**Reply by coltonehrman** [reply]
@brucesailes Join the discussion and pitch your ideas and feedback :) We are just in the planning phase right now, so mostly brainstorming

**Comment by tjay22** [comment]
## User Journey (Mentee)

1. Create mentee account (sign up with email, Google, etc.)
2. Mentee will be asked a series of questions to create their profile.
    - Name
    - Goals (they will choose from a provided list)
      - Learn something new
      - Master a skill
    - Experience
      - Less than 2 years
      - 2 - 5 years
      - More than 5 years
    - Mentee picks from a list of technologies they want to learn/develop. Note: The list will be updated based on the mentors available on the platform. If no mentors are available to teach Java for example, it won’t be listed.
      - Frontend
        - HTML, CSS, JavaScript, TypeScript (these are only examples)
      - Backend
        - Python, Java, Ruby, C#, PHP
      - Frameworks / Libraries
        - React, Angular, Vue, Svelte, Next.js
      - Mobile App Development
        - Swift, Android, Flutter, React Native
      - Focused Tech Stack
        - LAMP, MEAN, MERN, Python-Django
      - Others
        - AWS, Git, Writing Documentation, etc.
    - Ask Mentee if they prefer in-person or virtual sessions?
    - Location (optional) - to be used if Mentee prefers in-person sessions
3. Profile is created - add RUD (Read, Update, Delete) functionality.
4. Mentee lands on homepage that has list of mentors that match their criteria.
5. Mentee browses through the list. They can further refine/change their search criteria using the filter.
6. Mentee finds a suitable mentor and sends a request.
7. Mentor accepts the request.
8. Mentor sends Mentee their calendar (calendly, acuity, etc.) to book sessions.
9. If virtual, Mentor will send invites via Google Meet or Zoom<sup>1</sup>. If in-person, Mentor and Mentee will coordinate the location.
    - Alternatively, for in-person sessions, Mentor will have a predefined list of locations that the Mentee can pick from. If Mentee doesn’t agree with the location, they can request a virtual session.
  ***<sup>1</sup> We will use external video conferencing platforms for now until we decide whether to want to develop an internal video conferencing platform.***
10. After 5 sessions or so, Mentee and Mentor will be asked to give each other feedback.

## User Journey (Mentor)

1. Create mentor account (sign up with email, Google, etc.)
2. Mentor will be asked a series of questions to create their profile.
    - Name
    - Experience
      - Less than 2 years
      - 2 - 5 years
      - 5 - 10 years
      - More than 10 years
    - Mentor picks from a list of technologies they are confident in teaching.
      - Frontend
        - HTML, CSS, JavaScript, TypeScript (these are only examples)
      - Backend
        - Python, Java, Ruby, C#, PHP
      - Frameworks / Libraries
        - React, Angular, Vue, Svelte, Next.js
      - Mobile App Development
        - Swift, Android, Flutter, React Native
      - Focused Tech Stack
        - LAMP, MEAN, MERN, Python-Django
      - Others
        - AWS, Git, Writing Documentation, etc.
    - Ask Mentor if they prefer in-person or virtual sessions or both?
    - Location (optional) - to be used for in-person sessions
3. Profile is created - add RUD (Read, Update, Delete) functionality.
4. Mentor can add certifications, projects, links and experiences to their profile as well as add list of locations for in-person sessions.
5. Mentor turns on availability and waits to be matched with a Mentee.
6. When a request is made, Mentor accepts the request.
7. Mentor sends Mentee their calendar (calendly, acuity, etc.) to book sessions.
8. If virtual, Mentor will send invites via Google Meet or Zoom. If in-person, Mentor and Mentee will coordinate the location.
9. After 5 sessions or so, Mentee and Mentor will be asked to give each other feedback.

**Reply by coltonehrman** [reply]
@tjay22 

Questions for "account creation" flows.

1. Should we have separate pages/routes for mentor vs mentee account creation?
    - This could further customize the different experiences and also create a hard distinction between each type of member/user.
1. Are the questions required or optional for a user to begin using the app?
1. At what point do we show the series of questions to the user?
    - Before or after account creation?
    - Within the account creation flow?
    - Async  within the account dashboard after a user has already signed up and created the account?
1. Depending on when we show the questions, what is the best UX for the questions presented?
    - Type
        1. Pop-up/modal  
        1. Full-page
        2. Inline form
    - Inputs
        1. Long form view that scrolls
        2. Step by step form with single input
        3. Step by step form with groups inputs

**Reply by tjay22** [reply]
@coltonehrman 

1. I think separate routes is the way to go here because I think scalability is something we need to think about as well. By having that hard distinction, it will make it easier to maintain (I think - correct me if I'm wrong)
2. The users will have the option to skip the questions and manually update their profile later. Upon signup however, their email address will be saved, user ID created and a random user name will be assigned to them. For example, John Doe.
3. Within the account creation flow. This flow will only show up for new account creation. For those who choose to skip it, they can update their profiles in their Account Dashboard.
4. To stick with modern design, I think full-page type and step by step form with single inputs are the way to go

Of course, these are just my opinions based on my experiences so I am open to other ideas that the other team members have.

**Reply by coltonehrman** [reply]
@tjay22

1. I love that you are thinking about maintainability/scalability, but I wouldn't let those play a role in deciding how the product should be. We should just consider what makes the most sense for the product for our "customers" and then we can come up with the best possible engineered solution for that. There will always be trade-offs & compromises when it comes to engineering and product requirements lol. That being said, starting out with a combined route would actually make things simpler and faster to iterate on. We can always decide later to split off. Splitting later would be easier vs trying to combine later if we decided to start off split then combine down the road.
2. I agree! So, we are saying we will present the questions during signup, but they will be optional and presented again later?
3. Yes! I think this states what I just asked in line 2 lol
4. Nice! I will trust your expertise in UX here :D

**Reply by brucesailes** [reply]
Hello guys! I'm in here

**Reply by brucesailes** [reply]
What do I need to work on?

**Reply by coltonehrman** [reply]
Hey @brucesailes, not sure if you read through our discussion. I would suggest reading through and posting your own thoughts/ideas related to the project, and also providing feedback/questions/replies to the ongoing discussion we are having right now.

If you are eager to pickup some work, you can find something under https://github.com/Software-Developer-Academy/mentor-match/issues or inside the https://github.com/orgs/Software-Developer-Academy/projects/1.

Let me know if you have any questions about anything!

We are doing daily standup sessions in the Discord server as well if you are ever free to join in and listen and participate :)
Discord invite link - https://discord.gg/8zpKC3My

**Reply by tjay22** [reply]
@coltonehrman 

Thank you for explaining that. It also reiterates the agile methodology. Love it!

**Reply by coltonehrman** [reply]
> @coltonehrman 
> 
> 
> 
> Thank you for explaining that. It also reiterates the agile methodology. Love it!

@tjay22 I think we have enough detail to create a new story/issue related to the questionnaire form for new accounts. Wondering if you wanted to help me out by created that with the details we mentioned here?

**Reply by tjay22** [reply]
@coltonehrman 

For sure! I am just putting together a quick wireframe for the rest of the team so that we are all on the same page on what we are creating.

**Reply by coltonehrman** [reply]
@tjay22 Awesome! I look forward to see everything 👀

**Reply by hinode-codes** [reply]
Based from this, should we add a "what is your role?" options or something after a user signs up?

**Comment by mikebashford** [comment]
Just a thought, I think this was brought up in the initial meeting. Did we want to use something like clerk to handle user authentication?

**Reply by coltonehrman** [reply]
@mikebashford That's a good question. I personally think we should stay away from these as much as we can for now. Just to allow for more learnings/experience of integrating these type of things from scratch. Also, Clerk, although I love their service, adds some additional overhead to our development experience, since it requires webhooks, and enabling a webhook locally would require another tool.

**Reply by mikebashford** [reply]
This is very true, now that I remember we did mention using JWT.

**Comment by tjay22** [comment]
## Wireframe

Here is the wireframe. This will help visualize the stuff we've been discussing and how things will work together. I'd love to hear others thoughts on it. 

[Mentor Match.pdf](https://github.com/Software-Developer-Academy/mentor-match/files/14951979/Mentor.Match.pdf)

**Reply by mikebashford** [reply]
Looks awesome, easy to follow, good flow.

**Reply by brucesailes** [reply]
Looks good!

**Comment by hinode-codes** [comment]
### Realtime communication

Since this is a mentor-match platform, will we implement realtime communication for chats, or just display the mentor's social links, and meeting links or something? What I can think of is we use Socket IO since SSE is difficult on NextJS.

**Reply by hinode-codes** [reply]
This will be a separate big part of the project though, so we need to discuss this first. I know a fair amount of Socket IO, so I recommend it.
